File: leetcode_other/Trees/binary_tree_right_side_view.py
'''
https://leetcode.com/problems/binary-tree-right-side-view/

Given the root of a binary tree, imagine yourself standing on the right side of it, return the values of the nodes you can see ordered from top to bottom.

Input: root = [1,2,3,null,5,null,4]
Output: [1,3,4]

Input: root = [1,null,3]
Output: [1,3]

Input: root = []
Output: []

(An additional case)
Input: root = [1,2]
output: [1,2]

explanation of above case:
by right side view, we define this as the rightmost node of each level of the tree
that explains why a right side view of [1,2] is [1,2], despite 2 being the left child of 1
because in this case for the 2nd level, 2 is the only node that isn't None so that would be the rightmost child

Concept:
perform a level order traversal on our tree using a queue and BFS:
for each node that gets popped from the queue, add its left and right children to the queue
except on each level, we only add the last node of our queue to our result list, as this will be the rightmost node on our current level
(i.e the right most node that is visible if we were to be standing on the right side of the tree as mentioned in the problem description)

Time complexity: O(N), representing the nodes in the tree
space complexity: O(N), N nodes in the queue

Revisited on 8/9/2023
One important note is that we save the length of the queue before we start popping out elements from the level
so we know which element was last in the level

'''
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if (not root):
            return []
        queue = deque()
        queue.append(root)
        res = []
        while (queue):
            N = len(queue)
            for i in range(len(queue)):
                node = queue.popleft()
                ## the last item in the queue will be the rightmost node in our current level
                if i == N-1:
                    res.append(node.val)              
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
        return res
               
                
        

File: node_js_ds_and_algos_implementations/trees/binary_tree_right_side_view.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number[]}
     */
    rightSideView(root) {
        /*
        level order traversal using BFS
        but at each level, only take the very last element into the result 

        More optimal solution that uses less memory per level:
        just save the very last element in a variable, so by default, we overwrite the variable
        within each level and only the very last iteration gets saved, that becomes the right side
        */
        if (!root){
            return []
        }
        let q = []
        q.push(root)
        let res = []
        while (q.length){
            let N = q.length
            let rightSide = null
            for (let i = 0; i < N; ++i){
                let cur = q.shift()
                if (cur.left){
                    q.push(cur.left)
                }
                if (cur.right){
                    q.push(cur.right)
                }
                rightSide = cur.val
            }
            res.push(rightSide)
        }
        return res
    }
}

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
/*
Same as level order traversal,
but:
right side of the tree is always the last node
of each level within our level order traversal

After traversal is done, pull the last element out of each list
*/
var rightSideView = function(root) {
    if (!root){
        return []
    }
    let q = []
    let levels = []
    q.push(root)
    while (q.length){
        let originalLength = q.length
        let level = []

        for (let i = 0; i < originalLength; ++i){
            let node = q.shift()
            if (node.left){
                q.push(node.left)
            }
            if (node.right){
                q.push(node.right)
            }
            level.push(node.val)
        }
        levels.push(level)
    }
    return levels.map((level) => level[level.length-1])
};